src/sparselm/model/_composite.py

```diff
@@ -0,0 +1,179 @@
+"""Composite estimator for piece-wise fitting."""
+
+__author__ = "Fengyu Xie"
+
+import re
+from itertools import chain
+
+import numpy as np
+from numpy.typing import ArrayLike
+from sklearn.base import RegressorMixin
+from sklearn.linear_model._base import LinearModel
+
+
+class CompositeEstimator(RegressorMixin, LinearModel):
+    """A composite estimator used to do piece-wise fitting.
+
+    The first estimator in the composite will be used to fit
+    certain features (a piece of the feature matrix, called scope)
+    to the target vector, and the residuals are fitted to the rest
+    of features by using the next estimators in the composite.
+    """
+
+    def __init__(self, estimators, estimator_scopes):
+        """Initialize estimator.
+
+        Args:
+            estimators(list[CVXEstimator]):
+                A list of CVXEstimators to add. CompositeEstimator
+                cannot be a member of CompositeEstimator.
+                An estimator will fit the residuals leftover by
+                the previous estimator fits in the list.
+            estimator_scopes(list[list[int]]):
+                Scope of each estimator, which means the indices of
+                features in the scope (features[:, scope]) will be
+                fitted to the residual using the corresponding estimator.
+                Notice:
+                   If estimators in the composite requires hierarchy
+                   or groups, the indices in the groups or hierarchy
+                   must be adjusted such that they correspond to the groups
+                   or hierarchy relations in the part of features sliced
+                   by scope.
+                   For example, consider original groups = [0, 1, 1, 2, 2],
+                   and an estimator has scope = [3, 4], then the estimator
+                   should be initialized with group = [0, 0].
+                   You are fully responsible to initialize the estimators
+                   with correct hierarchy, groups and other parameters before
+                   wrapping them up with the composite!
+        """
+        if len(estimators) != len(estimator_scopes):
+            raise ValueError("Must specify a scope for each estimator!")
+        full_scope = sorted(set(chain(*estimator_scopes)))
+        if full_scope != sorted(chain(*estimator_scopes)) or full_scope != list(
+            range(len(full_scope))
+        ):
+            raise ValueError(
+                f"Estimator scopes: {estimator_scopes}"
```

Comment:
Rephrase as what is not allowed.  "scopes can not overlap and must be continuous"

Reply:
Done.